[MRG][DOC] Issue #14312 - Ensure All Attributes are Documented: Bayesian Ridge

juliapiscioniere opened the pull request:
Added Bayesian Ridge Regression attribute documentation for  X_offset_ , X_scale_ , and y_offset

<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs

Issue #14312
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.
This adds documentation for attributes for the Bayesian Ridge function. The attributes are x_offset, y_offset, and x_scale.

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

Commit:
[DOC] Issue #14312 - Ensure All Attributes are Documented
Added Bayesian Ridge Regression attribute documentation for  X_offset_ , X_scale_ , and y_offset

juliapiscioniere commented:
Hi! I am wondering if you would be able to take a look at this pull request? @jnothman

adrinjalali commented:
I'll probably take a look a bit later, for now, you have linter issues (line too long), which make the CI fail.

NicolasHug commented:
Thanks for opening a PR @juliapiscioniere .

Unfortunately I think these attributes are false positives (BTW did you reference the right issue?) that should have been private in the first place, ie with a leading `_` instead of a trailing one.

Would you be intersted in deprecating the parameters? We have instructions at https://scikit-learn.org/stable/developers/contributing.html#deprecation

Commit:
updating my version of scikit
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into julia_feature_branch

juliapiscioniere commented:
@NicolasHug Thank you so much for taking a look! The issue we are looking at is #14312- Ensure All Attributes are Documented. Can you please clarify what you are referencing with the leading versus trailing underscore? Most of the attributes on the issue's list have a trailing underscore, so is this specific to these attributes? We have started working on others and plan to submit PR's for those as well, so we want to make sure we are documenting attributes that need it. Also, maybe a trivial question, but why do these attributes need deprecation?

NicolasHug commented:
Estimators have attributes, some of them are public and some of them are private. Ideally, the attributes that might be useful to users are public, and those that are mostly internals of the algorithm should be private.

Public attributes have a trailing underscore, i.e. `self.public_attr_` the private ones have a leading underscore, i.e. `self._private_attr`.

Users are free to use `self.public_attr_` and we guarentee that we won't break backward compatibility on these attributes (for at least 2 major versions). However `self.public_attr_` might change without notice. When we decide to change a public attribute into a private attribute, we need a deprecation cycle which says "careful there, you're using a public attribute but we're making it private so you should not use it anymore". You'll find more details in the link provided above.

Currently,  `X_offset_` , `X_scale_` , and `y_offset_` are de facto public because they have trailing underscore. That's why they show up in the issue that you mentioned. However, my opinion is that they're not directly useful to the users, and they are mostly internals and implementation details that should be hidden. That's only my opinion and others might think otherwise.

So I would suggest to transform these public attributes into private ones, which would require a deprecation cycle.

(Maybe wait for the opinion of @adrinjalali and @jnothman )

adrinjalali commented:
> However self.public_attr_ might change without notice.

I think you meant "However `self._private_attr` might change without notice."

This is where I'd really love to have an easy way to check public repos and see how often people are using them lol. Personally, I agree with you @NicolasHug that they should be private. But that may be due to my ignorance of how people would be using those attributes.

jnothman commented:
I agree that these are largely implementation details, although they may be
useful for users trying to reproduce some of the results here. I'm okay to
document them and avoid the hassle to us and to the user of deprecation.
But the docs shouldn't mention private functions and should be more framed
in terms of how the user may use them together with other attributes.

Commit:
Shortening the line lengths of the documentation.

Commit:
Excaping the forward slash.

Commit:
Figuring out the linting issue - only adding one attribute

Commit:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into julia_feature_branch

Commit:
Adding the other two attributes back into the file, getting rid of the forward slash.

juliapiscioniere commented:
@jnothman @adrinjalali @NicolasHug Thanks all for the discussion. I fixed the linter issue (and am unsure if the other checks need to be fixed). Is there a final consensus on whether or not we should continue with documenting these attributes or open an issue for deprecation and continue with that for these attributes?

NicolasHug commented:
> I'm okay to document them and avoid the hassle to us and to the user of deprecation.

@jnothman a small clarification, are you suggesting to keep the attributes public, document them, but still treat them as private i.e. not go through deprecation cycles when/if we decide to change them on day?

jnothman commented:
I wouldn't treat them as private. I don't really think they're important
enough to bother removing, though.

jnothman commented:
I'm basically saying: choose a low effort solution.

juliapiscioniere commented:
Hi, I know that this is not a very important thing going on right now, but these pull requests are for classwork. If we would be able to get a response on this pull request by April 5th that would be awesome! Thank you for your responses so far, we understand if you don't have time.

cmarmo commented:
Hi @juliapiscioniere your PR has linting issues: you can check them locally using
```
flake8 sklearn/linear_model/_bayes.py
```
The output will be
```
sklearn/linear_model/_bayes.py:15:1: F401 '..utils.check_X_y' imported but unused
sklearn/linear_model/_bayes.py:122:48: W291 trailing whitespace
sklearn/linear_model/_bayes.py:125:1: W293 blank line contains whitespace
sklearn/linear_model/_bayes.py:126:35: W291 trailing whitespace
sklearn/linear_model/_bayes.py:130:48: W291 trailing whitespace
sklearn/linear_model/_bayes.py:132:18: W291 trailing whitespace
```
The first line is not linked to your modifications, but the error is not in master anymore... 
May I suggest to sync your master with upstream/master, then fix the remaining lint errors? Then we will ping the reviewers as you can finalize this by April 5th... is that ok for you? Thanks for your patience!

Commit:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into julia_feature_branch

Commit:
fixing whitespace linter issues

cmarmo commented on sklearn/linear_model/_bayes.py:
```diff
@@ -118,6 +118,18 @@ class BayesianRidge(RegressorMixin, LinearModel):
+    X_offset_ : array-like (length = n_samples)
+        The eventually weighted mean of regressors X, used to center X in
+        the _pre_process_data function.
+    
```
```suggestion

```
Some whitespaces still there...

cmarmo commented:
@adrinjalali , @jnothman , @NicolasHug, maybe the discussion about making those attributes private could be postponed and this PR reviewed? I think that we all felt classwork anxiety in our life... this cries for empathy... ;) ... Thanks for your time!

adrinjalali commented on sklearn/linear_model/_bayes.py:
```diff
@@ -118,6 +118,18 @@ class BayesianRidge(RegressorMixin, LinearModel):
 
+    X_offset_ : array-like (length = n_samples)
+        The eventually weighted mean of regressors X, used to center X in
+        the _pre_process_data function.
```
we shouldn't point to private methods in the public docs. In this case `_pre_process_data` is private, and in general anything starting with `_` (conventionally). Instead, you can try to find an explanation from the way these attributes are used. In this case, you could look at the `_set_intercept` to get an idea of how they work.

adrinjalali commented on sklearn/linear_model/_bayes.py:
```diff
@@ -118,6 +118,18 @@ class BayesianRidge(RegressorMixin, LinearModel):
     n_iter_ : int
         The actual number of iterations to reach the stopping criterion.
 
+    X_offset_ : array-like (length = n_samples)
```
you can see how to write these lines on https://scikit-learn.org/dev/developers/contributing.html#guidelines-for-writing-documentation

adrinjalali reviewed (commented):
I'd agree with @jnothman that it's not too important to make these attributes private. I'd leave them as is.

Commit:
Getting rid of whitespace in line 124`

cmarmo commented on sklearn/linear_model/_bayes.py:
```diff
@@ -118,6 +118,18 @@ class BayesianRidge(RegressorMixin, LinearModel):
+
+    X_scale_ : float, default = 64
+        L2-norm of X utilized in equation self.normalize:
+        X = (X - self.X_offset_) divided by (self.X_scale_)
```
```suggestion
        `X = (X - self.X_offset_) / (self.X_scale_)`
```
This will fix the Sphinx warning.

Commit:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into julia_feature_branch

Commit:
Fixing line 127

cmarmo commented:
Closing this pull request as documentation of the attributes `X_offset_` and `X_scale_` for `BayesianRidge` has been added with #18607 .
